### ISSUE TITLE
Idea Contribute: a way to cut off generated codes to make it more ecs-alike.

### ISSUE THREAD

**Issue by Kiddinglife** [sschmid/Entitas-CSharp#788]
Hi entitas,

Reason:
generated wrapper still put everything together which violates the basic principle of ecs.

Possible improvement:
1. Inheritage ScriptObject and use it as entity container that contains user-defined components.
2. Inheritage MonoBehaviour and we can use it as system and also can append it to game object from unity editor:
   Use it as InitialSystem when we write init codes inside start();
   Use it as IExucuteSystem when we write game logic codes inside update();
   Use it as ReactiveSystem by overriding unity build-in on***(), eg onCollisionEnter();

Regards.

**Comment by sschmid** [comment]
Hi, I don't quite understand what you're suggesting. Can you rephrase please?

**Comment by Kiddinglife** [comment]
Hi sschmid,
I write some pseudo codes that can show you my idea clearly.
Regards.

```c#
struct MonoBehavior
{
    void start()
    {
        
    }
    void update()
    {

    }
    void end()
    {
        
    }
}

struct sys
{
    MonoBehavior[] monoBehaviors;
    void add(MonoBehavior entity)
    {
        monoBehaviors.push(entity);
    }
    void remove(MonoBehavior entity)
    {
        monoBehaviors.push(entity);
    }
    void clear()
    {
        monoBehaviors.clear();
    }
}

struct component_move
{
    transform trans;
    animation walk;
    animation run;
    int speed;
    nevigation nag;
}
struct move : sys
{
    void update()
    {
        for(MonoBehavior entity : monoBehaviors)
        {
            if(dest_trans != null)
                entity.nag.moveto(dest_trans);
            if(walk != null)
            {
                speed = 10;
                entity.walk.play();
            }
            else
            {
                speed = 50;
                entity.run.play();
            }
            entity.trans += speed * time_interval;
            entity.nag.moveto(entity.trans);
        }
    }
}

struct component_health
{
    int health;
    bool alive;
    animation dead;
    animation alive;
}
struct on_health_change : sys
{
    void update()
    {
        for(MonoBehavior entity : monoBehaviors)
        {
            entity.health -= entity.shell;
            if(entity.health == 0)
            {
                alive = false;
                entity.dead.play();
            }
        }
        this.clear();
    }
}

struct component_collision
{
    int shell;
    animation animation_hit;
}
struct on_collision : sys
{
    void update()
    {
        for(MonoBehavior entity : monoBehaviors)
        {
            entity.shell -= 1;
            entity.move.speed = -100; // simulate to bounce back
            entity.animation_hit.play();
        }
        // event-triggered system should always be removed after used
        this.clear();
    }
}

class application : MonoBehavior
{
    sys[][] sys = new array[8][8];

    override void start()
    { 
        sys[0].add(new on_health_change());
        sys[0].add(new move());
        sys[1].add(new on_collision_enter());
    }

    override void update()
    {
        for(sys s : update_sys)
            s.update();
    }

    overide void FixedUpdate()
    {
        for(sys s : fix_update_sys)
            s.update();
    }

    sys get(int systype, int sysidx)
    {
        return sys[systype][sysidx];
    }
}

class mob_entity : MonoBehavior
{
    MonoBehavior application;

    component_collision collision;
    component_health health;
    component_move move;

    override void start()
    { 
        application = gmeobject.find('application');
        
        // Append position_son_position_change to this entity inside start() method 
        // because position component is constantly being updated in each frame 
        move_g.add(this);
    }

    override void onCollision()
    {
        // Append on_collision and on_health_change to this entity 
        // when another entity hits us told by this entity's rigibody system that calls onCollisionEnter().
        // on_collision and on_health_change are regarded as event-triggered system or one-off system 
        // they will be removed from the entity after used
        sys on_collision = applicationon.get(1,0);
        sys on_health_change = applicationon.get(0,0);
        on_collisionget.add(this);
        on_health_change.add(this);
    }
}
```
